3 :: View to edit only favourite currency

JonhyOliveira opened the pull request:
Previously the user had to edit his registration info to change his favourite currency, which flooded him with unnecessary information which contributed to a worse UX.

Now, in a separate view, the user can edit **only** his favourite currency without the need for his current password.

JonhyOliveira committed:
implemented feature

JonhyOliveira commented:
This branch was developed against #3

gdiasdasilva commented on rails_app/app/controllers/currencies_controller.rb:
```diff
@@ -5,7 +5,7 @@
 class CurrenciesController < ApplicationController
 
-  before_action :authenticate_user!, only: [:follow, :unfollow, :followed]
+  before_action :authenticate_user!, only: [:follow, :unfollow, :followed, :change_favorite, :edit_favorite]
```
Since almost all of the actions rely on this callback, it would be simpler to define it using `except` instead of `only.`

JonhyOliveira replied:
:+1:

JonhyOliveira replied:
Fixed in 84b08a9d3896aa7fc9b7ced2c8cf1d144ca52210

gdiasdasilva commented on rails_app/app/views/currencies/edit_favorite.html.erb:
```diff
@@ -0,0 +1,18 @@
+
+  <h1>Edit Favorite Currency</h1>
+
+  <%= form_for(@user, url: change_favorite_currency_path, html: { method: :put }) do |f| %>
```
Couldn't this be `current_user` directly without the need of a new variable?

JonhyOliveira replied:
:+1:

JonhyOliveira replied:
Fixed in 43ae180e6b56cbe31bc574671d79e22648628a15

gdiasdasilva commented on rails_app/app/controllers/currencies_controller.rb:
```diff
@@ -41,6 +41,16 @@ def all
+  end
+
+  def change_favorite
+    if current_user.update favorite_currency_id: params[:user][:favorite_currency_id]
```
Since in #3 now you've set `favorite_currency` has an association in the `User` model, this could be done in a different way:

```ruby
currency = Currency.find(id: params[:user][:favorite_currency_id])
current_user.update(favorite_currency: currency)

gdiasdasilva replied:
This way, if a unrecognised currency id is submitted, it will fail.

JonhyOliveira replied:
:+1:

JonhyOliveira replied:
Adressed in d4fdea0ca5a96a4648e1e08fe6f7f760212c8d2f

gdiasdasilva reviewed (commented):
👀

JonhyOliveira committed:
Fixing readability

Addressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/4/files#r1179239843

JonhyOliveira committed:
Making change_favorite easier to read

Addressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/4/files#r1179246067

JonhyOliveira committed:
View is less confusing

Adressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/4/files#r1179241163

gdiasdasilva commented on rails_app/app/controllers/currencies_controller.rb:
```diff
@@ -15,6 +15,18 @@ def index
+    if user_signed_in?
+      user = current_user
+
+      @followed_codes = user.followed_currencies.map { |curr| curr.code }
```
This is iterating all currencies in memory and then fetching the code of each one. I believe it could be done directly with:

```ruby
user.followed_currencies.pluck(:code)
```

Give it a try!

Reference: https://apidock.com/rails/ActiveRecord/Calculations/pluck

JonhyOliveira replied:
:+1:

JonhyOliveira replied:
Addressed in 4883663275ee1e01c41a98ed576184212e83889f
Yep, all db-side now <img width="753" alt="Screenshot 2023-04-28 at 16 20 37" src="https://user-images.githubusercontent.com/54363138/235187948-63b20ba2-58d9-4b34-9426-a745b1a953b0.png">

JonhyOliveira committed:
Currency code query now uses `pluck`

Addressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/4#discussion_r1180224552
